### PR TITLE
Feature/merge tenants

### PR DESCRIPTION
**Pull request by vzickner**
* adding hooks for pre-merging and finalize-merging (before clearing local cache)
* default implementations for hooks to do a merge (see included documentation for more details)

#### Check List:
* Unit tests: YES
* Documentation: YES

### CODE REVIEW

**Comment by dbmalkovsky** [comment]
@vzickner Why is only the Chinese documentation updated? 
See `docs/userguide/src/zh_CN/bpmn/ch18-Advanced.adoc`

File: modules/flowable-engine/src/main/java/org/flowable/engine/impl/cmd/ChangeDeploymentTenantIdCmd.java [hunk 1]
```diff
@@ -15,12 +15,19 @@
...
 import org.flowable.common.engine.impl.interceptor.Command;
 import org.flowable.common.engine.impl.interceptor.CommandContext;
 import org.flowable.engine.impl.ProcessDefinitionQueryImpl;
 import org.flowable.engine.impl.persistence.entity.DeploymentEntity;
+import org.flowable.engine.impl.repository.AddAsNewDeploymentMergeStrategy;
+import org.flowable.engine.impl.repository.AddAsOldDeploymentMergeStrategy;
+import org.flowable.engine.repository.DeploymentMergeStrategy;
+import org.flowable.engine.impl.repository.MergeByDateDeploymentMergeStrategy;
+import org.flowable.engine.repository.MergeMode;
+import org.flowable.engine.impl.repository.VerifyDeploymentMergeStrategy;
```
Comment (dbmalkovsky) [inline modules/flowable-engine/src/main/java/org/flowable/engine/impl/cmd/ChangeDeploymentTenantIdCmd.java]: It appears that the imports are not in alphabetic order.

File: modules/flowable-engine/src/main/java/org/flowable/engine/impl/repository/AddAsNewDeploymentMergeStrategy.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package org.flowable.engine.impl.repository;
```
Comment (dbmalkovsky) [inline modules/flowable-engine/src/main/java/org/flowable/engine/impl/repository/AddAsNewDeploymentMergeStrategy.java]: Missing Apache license header.

File: modules/flowable-engine/src/main/java/org/flowable/engine/impl/repository/AddAsOldDeploymentMergeStrategy.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
+package org.flowable.engine.impl.repository;
```
Comment (dbmalkovsky) [inline modules/flowable-engine/src/main/java/org/flowable/engine/impl/repository/AddAsOldDeploymentMergeStrategy.java]: Missing license header.

File: modules/flowable-engine/src/test/java/org/flowable/engine/test/api/tenant/TenancyTest.java [hunk 4]
```diff
@@ -12,16 +12,19 @@
...
 import java.util.ArrayList;
 import java.util.List;
 
+import org.apache.ibatis.exceptions.PersistenceException;
 import org.flowable.common.engine.api.FlowableException;
 import org.flowable.common.engine.impl.history.HistoryLevel;
 import org.flowable.common.engine.impl.util.CollectionUtil;
 import org.flowable.engine.ProcessEngineConfiguration;
 import org.flowable.engine.history.HistoricActivityInstance;
+import org.flowable.engine.repository.MergeMode;
```
Comment (dbmalkovsky) [inline modules/flowable-engine/src/test/java/org/flowable/engine/test/api/tenant/TenancyTest.java]: Not alphabetic order here too.

**Comment by vzickner** [comment]
Thanks for the review @dbmalkovsky. I just pushed an update.

File: modules/flowable-engine/src/main/java/org/flowable/engine/repository/MergeMode.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.flowable.engine.repository;
+
+/**
+ * @author Valentin Zickner
+ */
+public class MergeMode {
```
Comment (PascalSchumacher) [inline modules/flowable-engine/src/main/java/org/flowable/engine/repository/MergeMode.java]: Maybe this should be an `enum`? What do you think?

Comment (vzickner) [reply modules/flowable-engine/src/main/java/org/flowable/engine/repository/MergeMode.java]: I changed it to the `enum`. Initially the idea was that people are able to extend the list and that was the reason for using a `String`. At the end I picked the approach where they can simply provide their own `DeploymentMergeStrategy` as a parameter.

**Pull request merged**